src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpContentReadStream.cs

```diff
@@ -46,7 +46,8 @@ protected override void Dispose(bool disposing)
                     return;
                 }
 
-                if (disposing && NeedsDrain)
+                // The connection might be disposed because of cancellation. We should not drain the response in this case.
+                if (disposing && NeedsDrain && _connection?._disposed != Status_Disposed)
```

Comment:
If _connection is null, it will still pass this check.  Is that intended?

Also, is there something somewhere that prevents this transition immediately after this check happens?

Reply:
> If _connection is null, it will still pass this check. Is that intended?

No, I will fix this. (Although it doesn't cause any problems since `_connection == null` leads to `NeedsDrain == false`.)

> Also, is there something somewhere that prevents this transition immediately after this check happens?

In the race I described in https://github.com/dotnet/runtime/issues/56159#issuecomment-897680702 connection should be always disposed at this point, or won't be disposed later. I'm not aware of other cases. @geoffkizer thoughts?

Reply:
> Also, is there something somewhere that prevents this transition immediately after this check happens?

Active prevention - I don't think so. But I also don't see any such transition in the code ATM.

Would a similar check for `Status_Disposed` help if it was added right before we return the connection to the pool? We're already checking for connection close there, so it might make sense...

Reply:
I moved the `_connection` checking to `NeedsDrain` implementations in 9e8f5ba, I think the code looks more straightforward in the new form.

Regarding the second concern: if connection disposal could somehow kick in immediately after the disposal of `HttpContentReadStream` that could cause a problem. I don't see how to construct such a scenario, even if it's possible, it seems "corner-casish" to me.

The PR fixes a serious problem with a common `Send(request, HttpCompletionOption.ResponseContentRead, cancellationToken)` case as-is.